### PR TITLE
MARKET-1546 Ace Editor

### PR DESCRIPTION
**Pull request by Saharij**
UI Builder - https://develop.backendless.com/app/zahar_mychka_components/ui-builder/default/pages/landing
Desktop Preview - https://earlywrench.backendless.app/api/files/ui-builder/containers/default/index.html?page=landing
Screenshot 2023-01-24 at 01 40 59
Screenshot 2023-01-24 at 01 45 26
Screenshot 2023-01-24 at 01 51 10

### CODE REVIEW

File: components/bl-ace-editor-component/src/index.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  component.getValue = () => editorValue;
+  component.setValue = value => setEditorValue(value);
+
+  useEffect(() => {
+    setEditorValue(value);
+  }, [value]);
+
+  useEffect(() => {
+    component.el = rootRef.current;
+  }, [rootRef]);
```
Comment (StasivPavlo) [inline components/bl-ace-editor-component/src/index.js:28]: Now we don't need this. Do how on example
Example:
```
export default function CustomUIComponent({ component, elRef }) {

 if(!component.display){
   return null
 }
 
  return (
    <div ref={elRef}>
      Custom Component with EL Ref
    </div>
  )
}
```

**Review by StasivPavlo — CHANGES_REQUESTED** [review]

File: components/bl-ace-editor-component/src/index.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        highlightActiveLine={ highlightActiveLine }
+        tabSize={ tabSize }
+        showPrintMargin={ printMarginVisibility }
+        setOptions={ {
+          printMarginColumn,
+          showInvisibles,
+          enableLiveAutocompletion: autocompletion,
+          enableBasicAutocompletion: autocompletion,
+          foldStyle,
+        } }
```
Comment (v-excelsior) [inline components/bl-ace-editor-component/src/index.js:56]: Create options before passing them to component

File: components/bl-ace-editor-component/src/index.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  component.setValue = value => setEditorValue(value);
+
+  useEffect(() => {
+    setEditorValue(value);
+  }, [value]);
+
+  const handleValueChange = useCallback((newValue) => {
+    setEditorValue(newValue);
+    onChange({ value: newValue });
+  }, []);
```
Comment (v-excelsior) [inline components/bl-ace-editor-component/src/index.js:28]:
```suggestion
  const onChangeHandler = useCallback(value => {
    setEditorValue(value);
    onChange({ value });
  }, []);
```

File: components/bl-ace-editor-component/src/index.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+export default function AceEditorComponent({ component, elRef, eventHandlers }) {
+  const {
+    classList, display, style, disabled, name, value, mode, theme,
+    foldStyle, placeholder, width, height, fontSize, tabSize, printMarginColumn,
+    printMarginVisibility, gutterVisibility, autocompletion, highlightActiveLine, showInvisibles
+  } = component;
+  const { onChange } = eventHandlers;
+
+  const [editorValue, setEditorValue] = useState('');
```
Comment (v-excelsior) [inline components/bl-ace-editor-component/src/index.js:16]: `useState(value)` ?

File: components/bl-ace-editor-component/src/index.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+  return (
+    <div ref={ elRef } className={ cn("bl-customComponent-aceEditor", classList) } style={ style }>
+      <AceEditor
+        value={ editorValue }
+        mode={ mode }
+        theme={ theme }
+        width={ width }
+        height={ height }
+        readOnly={ disabled }
```
Comment (v-excelsior) [inline components/bl-ace-editor-component/src/index.js:42]: > The difference between disabled and readonly is that read-only controls can still function and are still focusable, whereas disabled controls can not receive focus and are not submitted with the form and generally do not function as controls until they are enabled.

Is it really disabled? May be we need to use same definition as library(readonly)?

**Review by v-excelsior — CHANGES_REQUESTED** [review]

**Pull request merged**